c10/core/impl/cow/simulator.h

```diff
@@ -0,0 +1,60 @@
+#pragma once
+
+#include <c10/macros/Macros.h>
+#include <c10/util/intrusive_ptr.h>
+
+#include <cstdint>
+#include <mutex>
+#include <type_traits>
+
+namespace c10::impl::cow {
+
+namespace detail {
+
+// An empty struct to use as the base class for a non-intrusive
+// SimulatorImpl.
+struct Empty {};
+
+} // namespace detail
+
+// Simulates a copy-on-write storage for a tensor.
+//
+// This simulator is used to identify situations where a copy-on-write
+// tensor would have been created and whether or not a logical copy
+// uses the storage after a different copy has modified it. So in
+// theory, it can identify reads/writes to copies after a write to a
+// different copy that is presently an alias.
+//
+// However, the fidelity of this check is limited by the extent to
+// which we have instrumented operations as reading or writing to
+// storage.
+//
+// We use a monotonically increasing generation number to track
+// modifications to storage.
+template <bool intrusive = true>
```

Comment:
Why do we template on intrusive?

Reply:
I was seeing ASan issues with stack allocated intrusive_ptr_target subclass. That is expected?

We don't need the default shadow storage in the cow::StateMachine to be refcounted, since its lifetime is bound to that instance.

Come to think about it, do we need to refcount any of them? We could just make them live as long as the storage lives, and just allocate them in a std::deque or set-like object that doesn't invalidate. It would be probably slight tradeoff of peak memory vs refcount bumps.

An upside of doing that is that we could enhance the assert in maybe bump to verify the association of each shadow storage with the physical storage.

Reply:
This probably doesn't matter. Our cost model is that reshape() is expensive, but everything else is cheap now.